src/main/solution_space/distance.py

```diff
@@ -1,44 +1,88 @@
+lock = Lock()
 
-Item = TypeVar('Item')
+Item = TypeVar('Item', bound=IdCounter)
 Upd = TypeVar('Upd')
 
 
 class IDistanceMatrix(Generic[Item, Upd], metaclass=ABCMeta):
+    _thread_number = 8
+
+    class DistFiller(Thread, IdCounter):
+        def __init__(self, queue: Queue, distance_matrix: IDistanceMatrix):
+            Thread.__init__(self)
+            IdCounter.__init__(self)
+            log.info(f'Init {self.__class__.__name__} {self.id}')
+            self._queue = queue
+            self._dist_matrix = distance_matrix
+
+        def run(self) -> None:
+            while True:
+                log.info(f'Run {self.__class__.__name__} {self.id}')
+                src_item, dst_item = self._queue.get()
+                try:
+                    # It seems we don't need to lock anything since dict is thread safe as well as GumTreeDiff calling
+                    dist = self._dist_matrix._IDistanceMatrix__find_dist(src_item, dst_item)
+                    log.info(f'Found dist for src {src_item.id}, dst {dst_item.id} is {dist}')
+                    self._dist_matrix._dist[src_item][dst_item] = dist
+                finally:
+                    log.info(f'Task done by {self.__class__.__name__} {self.id}')
+                    self._queue.task_done()
+
     def __init__(self, to_store_dist: bool = True):
-        # Todo: is it better to use id as a key instead of Item?
         self._dist: Dict[Item, Dict[Item, int]] = {}
         self._to_store_dist = to_store_dist
 
+    @property
+    def to_store_dist(self) -> bool:
+        return self._to_store_dist
+
+    @to_store_dist.setter
+    def to_store_dist(self, value: bool):
+        log.info(f'Change to_store_dist from {self._to_store_dist} to {value}')
+        # If we change to_store_dist from False to True, we need to fill dist dict
+        if not self._to_store_dist and value:
+            self.__fill_dist()
+        self._to_store_dist = value
+
     # If we have stored a distance between src_item and dst_item, we return it, otherwise we find it explicitly.
     def get_dist(self, src_item: Item, dst_item: Item) -> int:
         dist = self._dist.get(src_item, {}).get(dst_item, None)
-        if dist is None:
+        #  If we don't store dist, dict can contain incorrect distances, so we should explicitly find it
+        if not self._to_store_dist or dist is None:
             return self.__find_dist(src_item, dst_item)
         return dist
 
     def add_dist(self, new_item: Item) -> bool:
-        if not self._to_store_dist:
-            log.info('The param to_store_dist is False. We don\'t use distance matrix')
-            return False
-
         if new_item in self._dist.keys():
             log.info('This item already exists')
             return False
+
+        # If we don't store dist, we don't fill self._dist[new_item],
+        # but we add an empty dict to be able to fill self._dist later
         self._dist[new_item] = {}
+        if not self._to_store_dist:
```

Comment:
No, we shouldn't. We need to store an empty dict for each vertex because when we will fill dist_matrix we should have all vertices as keys. 
If we swap these lines, we won't be able to fill dist matrix.
I wrote a comment about it right above line 81.